### ISSUE TITLE
Allow MetricsTracker to be set/updated/overwritten after the pool was created?

### ISSUE THREAD

**Issue by slyoldfox** [brettwooldridge/HikariCP#279]
In follow-up of issue #112, it would be nice if the MetricsTracker could be set after that the BaseHikariPool() is constructed.

In our project (https://bitbucket.org/beforeach/across-standard-modules/src/0e73c0beb971bac0e363eddb49a337996790d7b4/metric-module/src/main/java/com/foreach/across/modules/metrics/config/DataSourceMetricModuleConfiguration.java?at=spring-data) we are writing a spring module which "injects" metrics if HikariDataSources are found. We would be autowiring the datasources and if they are HikariDataSource instances, set the MetricsTracker accordingly. However, we have a bit of a catch-22.

DataSources in the across framework are initialized very early in the bootstrap (because it applies liquibase installers and whatnot). It would be impossible for our MetricRegistry module to create a bean before that (because the modules are bootstrapped after datasources are bootstrapped).

To cut a long story short, it would be nice if the MetricsTracker could also be changed/updated after the pool was created (without having to expose the pool explicitly of  course)

Right now we have this hackerish code which seems to work "okay" as a test (but of course will fail "isRecordMetrics" checks in BaseHikariPool.getConnection()).

````java
        DirectFieldAccessor df =  new DirectFieldAccessor(hikariDataSource);
        Object pool = df.<HikariPool>getPropertyValue("pool");
        if( pool instanceof HikariPool ) {
            HikariPool hikariPool = (HikariPool) pool;
            DirectFieldAccessor poolAccessor = new DirectFieldAccessor(hikariPool);
            Object metricsTracker = poolAccessor.getPropertyValue("metricsTracker");
            if( metricsTracker instanceof MetricsTracker) {
                ((MetricsTracker) metricsTracker).close();
                CodaHaleMetricsTracker something = new CodaHaleMetricsTracker(hikariPool, new MetricRegistry() );
                poolAccessor.setPropertyValue( "metricsTracker", something );
            }
        }
        return hikariDataSource;
```
Is there any reason that we cannot update the MetricsTracker after the pool was created? Would you accept a PR for it? E.g. introduce a method updateMetricRegistry(MetricRegsitry) on HikariDataSource which delegate to the pool and overwrites the metricsTracker internally (of course closing() and instantantiating the correct fields where needed?

**Comment by brettwooldridge** [comment]
@slyoldfox Before I commit one way or another let me run some performance tests.  Generally speaking, HikariCP uses *final* fields wherever possible.  Doing so allows the JVM to perform optimizations that would not otherwise be possible.   Whether the JVM is able to detect and eliminate "dead code" (DCE) depends on a variety of factors that are somewhat opaque.

If you look at the [MetricsTracker](https://github.com/brettwooldridge/HikariCP/blob/a310df95d71c0e3248efb9322f6a62a19590402d/hikaricp-common/src/main/java/com/zaxxer/hikari/metrics/MetricsTracker.java) class, you see the default implementation has no method bodies.  Further, take a look at [``BaseHikariPool.getConnection()``](https://github.com/brettwooldridge/HikariCP/blob/f7df7cd0b0a6455c1678350bb5f14050fcaa0d47/hikaricp-common/src/main/java/com/zaxxer/hikari/pool/BaseHikariPool.java#L193), you'll see these lines of code:
```java
198 final MetricsContext metricsContext = (isRecordMetrics ?
                                           metricsTracker.recordConnectionRequest(start) :
                                           MetricsTracker.NO_CONTEXT);
...
213 metricsContext.setConnectionLastOpen(bagEntry, now);
214 metricsContext.stop();
``` 
When metrics are *not* enabled, the JVM is able to detect that all of those methods result in zero code execution and the JIT completely eliminates those call-sites .  If you run a microbenchmark with microsecond resolution, you can actually see that the performance of the code is the same when those lines are actually deleted from the code and when they are present but metrics are disabled.

If ``isRecordMetrics`` is mutable (*non-final*) it also must be *volatile* because of cross-thread visibility requirements.  Once ``isRecordMetrics`` is *volatile* the JVM is required to emit code with a load-fence memory barrier, which has a large performance impact.  Additionally, because ``metricsTracker`` must also become mutable (*non-final*) the JVM can no longer be sure that the code in ``getConnection()`` is truly dead-code (even when metrics are disabled).  The JIT is therefore required to emit code that invokes ``metricsContext.setConnectionLastOpen()`` and ``metricsContext.stop()`` even in the case when metrics are disabled.

Again, I haven't benchmarked the code, but my gut feeling after doing this level of optimization for several years is that there will be a fairly high runtime penalty for making those mutable (changeable at runtime).

**Comment by brettwooldridge** [comment]
@slyoldfox Well, some good news.  The benchmarks are showing an ~1.7% decrease in performance by making ``isRecordMetrics`` and ``metricsTracker`` mutable, which is entering the realm of acceptable.

I am, however, considering making ``metricsTracker`` a "set-once" property.  But I believe that would be acceptable for your use-case, is that correct?

**Comment by slyoldfox** [comment]
Hi Brett,

Thanks for taking the time to look into this. I indeed read about the removal of dead code in some other threads, I think it was in #112. Anyway a set-once property sounds fine, as long as it's able to set the metricsTracker after the pool was initialized and accessible from HikariDataSource (ideally). That way the metricsTracker could be easily set on @Bean initialization, independent of the Bean orders.

When you have a snapshot around, I'll gladly give it a try.

**Comment by brettwooldridge** [comment]
@slyoldfox HikariCP 2.3.4 has been published to the maven central repository, and contains this enhancement:
```
 * Allow Dropwizard MetricRegistry/HealthCheckRegistry to be set after pool startup -- one time 
   only.
```

**Comment by slyoldfox** [comment]
Tnx @brettwooldridge, I had already noticed and was in the process of bumping the dependencies in our framework.

I came across some dbcp 1.4 dependencies we use for unit testing/IT testing (using BasicDatasource). Although we don't use database access much (basically just installing test data and run some liquibase installer scripts), would you recommend using a (your) connection pool in this aspect? Or would you just limit it to more "heavier" production usage?

Also I saw that
 <dependency>
<groupId>javax.inject</groupId>
<artifactId>javax.inject</artifactId>
<version>1</version>
</dependency>
wasn't marked as a test scope, although it seems only used in tests, is this a mistake or required (just asking because I noticed the dependency being added in our framework as well now.

Thanks for the good work :)

**Comment by brettwooldridge** [comment]
Doh, just about 10 minutes late.  I just published 2.3.5.  I marked javax.inject as 'test', but that won't appear until the next release, unless you want to build yourself.

No reason not to use HikariCP in place of dbcp for test, fewer dependencies right?

**Comment by slyoldfox** [comment]
Exactly, one of the reasons I saw the javax.inject :)

I'll keep an eye out for 2.3.5, thanks!
